server/src/main/java/org/elasticsearch/transport/InboundHandler.java

```diff
@@ -111,17 +114,26 @@ private void messageReceived(TcpChannel channel, InboundMessage message) throws
                 }
                 // ignore if its null, the service logs it
                 if (handler != null) {
-                    if (header.isError()) {
-                        handlerResponseError(streamInput, handler);
+                    final StreamInput streamInput;
+                    if (message.getContentLength() > 0) {
+                        streamInput = namedWriteableStream(message.openOrGetStreamInput());
+                        assertRemoteVersion(streamInput, header.getVersion());
+                        if (header.isError()) {
+                            handlerResponseError(streamInput, handler);
+                        } else {
+                            handleResponse(remoteAddress, streamInput, handler);
+                        }
+                        // Check the entire message has been read
+                        final int nextByte = streamInput.read();
+                        // calling read() is useful to make sure the message is fully read, even if there is an EOS marker
+                        if (nextByte != -1) {
+                            throw new IllegalStateException("Message not fully read (response) for requestId ["
+                                + requestId + "], handler [" + handler + "], error [" + header.isError()
+                                + "]; resetting");
+                        }
                     } else {
-                        handleResponse(remoteAddress, streamInput, handler);
-                    }
-                    // Check the entire message has been read
-                    final int nextByte = streamInput.read();
-                    // calling read() is useful to make sure the message is fully read, even if there is an EOS marker
-                    if (nextByte != -1) {
-                        throw new IllegalStateException("Message not fully read (response) for requestId [" + requestId + "], handler ["
-                            + handler + "], error [" + header.isError() + "]; resetting");
+                        assert header.isError() == false;
+                        handleResponse(remoteAddress, EMPTY_STREAM_INPUT, handler);
```

Comment:
This may introduce a BWC problem in the case that a response is only empty in some versions, because `EMPTY_STREAM_INPUT#getVersion()` returns `Version.CURRENT` which may not match `header.getVersion()`.

Reply:
Yea that was a little lazy on my end, I couldn't find a case where we ever moved from an empty message to a non-empty one but that doesn't mean that won't happen in the future. I adjusted the logic accordingly to use an empty stream per thread.